File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/Book.java
package org.babyfish.jimmer.sql.model;

import org.babyfish.jimmer.sql.*;
import org.babyfish.jimmer.sql.meta.UUIDIdGenerator;
import org.jetbrains.annotations.Nullable;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.Positive;
import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

/**
 * The Book Entity $:)$
 */
@Entity
public interface Book {

    /**
     * Id $:)$
     */
    @Id
    @GeneratedValue(generatorType = UUIDIdGenerator.class)
    UUID id();

    /**
     * Name $:)$
     */
    @Key
    @NotEmpty(message = "The book name cannot be empty")
    String name();

    /**
     * Edition $:)$
     */
    @Key
    int edition();

    /**
     * Price $:)$
     */
    @Positive
    BigDecimal price();

    /**
     * Store $:)$
     *
     * <p>Note: This property can be null</p>
     */
    @Nullable
    @ManyToOne
    BookStore store();

    /**
     * Authors $:)$
     */
    @ManyToMany
    @JoinTable(
            name = "BOOK_AUTHOR_MAPPING",
            joinColumnName = "BOOK_ID",
            inverseJoinColumnName = "AUTHOR_ID"
    )
    List<Author> authors();

    /**
     * StoreId $:)$
     *
     * <p>Note: This property can be null</p>
     */
    @IdView
    @Nullable
    UUID storeId();

    /**
     * AuthorIds $:)$
     */
    @IdView("authors")
    List<UUID> authorIds();
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/classic/book/Book.kt
package org.babyfish.jimmer.sql.kt.model.classic.book

import org.babyfish.jimmer.sql.*
import org.babyfish.jimmer.sql.kt.model.classic.store.BookStore
import org.babyfish.jimmer.sql.kt.model.classic.author.Author
import java.math.BigDecimal
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.Positive
import javax.validation.constraints.PositiveOrZero

@Entity
interface Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    @Key
    @get:NotEmpty(message = "The book name cannot be empty")
    val name: String

    @Key
    val edition: @PositiveOrZero Int

    val price: @Positive BigDecimal

    @ManyToOne
    val store: BookStore?

    @ManyToMany
    val authors: List<Author>

    @IdView
    val storeId: Long?

    @IdView("authors")
    val authorIds: List<Long>
}